### PR TITLE
connect react app to contract

### PR DESCRIPTION
**Pull request by amhunt**
This gets the web app to a minimally working state with our smart contracts (wooo!). I tested using MetaMask in Chrome. This does NOT add a way to view your file in the browser, but I was able to verify that the submitted text was being saved correctly on my local text network using command line.

NB: The Blockbin contract address constant in `InputForm` needs to be updated to whatever the address is when you deploy it on your local testnet. Obviously once we deploy the contract on the actual network, we will replace it with that address, which should be constant.

Up next, need a way for the web app to display/retrieve saved text files (should it just be `blockbin.io/[content_hash]` or something more creative? Maybe an input form to retrieve your files on the main page?).

@JFickel @ArnaudBrousseau @n-white

### CODE REVIEW

File: blockbin/src/components/InputForm.js [hunk 1]
```diff
@@ -1,16 +1,73 @@
 import React, { Component } from 'react';
+import Web3 from 'web3';
 
 import '../App.css';
 
+const CONTRACT_ADDRESS = '0xa022cc6ff8895fd07b03c480d46f20c412716fdc';
```
Comment (ArnaudBrousseau) [inline blockbin/src/components/InputForm.js]: This seems like something that should be put in configs? I'm pretty sure that we'll all get different addresses on our test networks, and once we deploy this "for real" we'll have one "prod" contract that's again different.

File: blockbin/src/components/InputForm.js [hunk 2]
```diff
@@ -1,16 +1,73 @@
...
+    };
+
+    if (typeof window.web3 !== 'undefined') {
+      console.log('using mist or metamask');
+      const newWeb3 = new Web3(window.web3.currentProvider);
+      this.web3 = newWeb3;
+    } else {
+      console.log('why aren\'t you using mist or metamask?!');
+      // set the provider you want from Web3.providers
+      this.web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));
```
Comment (ArnaudBrousseau) [inline blockbin/src/components/InputForm.js]: what is this for? Do we want to support non-metamask users? As a first step, seems fine to just do `window.alert("you are not using metamask, please install it and reload this page")`

File: blockbin/src/components/InputForm.js [hunk 3]
```diff
@@ -1,16 +1,73 @@
...
+    }
+    // Replace with correct ABI
```
Comment (ArnaudBrousseau) [inline blockbin/src/components/InputForm.js]: This should probably live in a separate file? Feels weird to have this in the InputForm component :)

(but no big deal, we can split later too)

File: blockbin/src/components/InputForm.js [hunk 4]
```diff
@@ -1,16 +1,73 @@
...
+    const cubeBytes = this.web3.fromAscii(this.state.pastedText);
+    const contentHash = this.web3.sha3(cubeBytes);
+
+    // metamask uses accounts[0] to pass the preferred account
+    const txId = this.contractInstance.dumpCube.sendTransaction(
+      cubeBytes, 
+      contentHash, 
+      {
+        from: this.web3.eth.accounts[0], 
+        gas: 2000000,
```
Comment (ArnaudBrousseau) [inline blockbin/src/components/InputForm.js]: Can this be configurable?

File: blockbin/src/components/InputForm.js [hunk 5]
```diff
@@ -1,16 +1,73 @@
...
+      },
+      function(error, result){
+        if(!error)
```
Comment (ArnaudBrousseau) [inline blockbin/src/components/InputForm.js]: 2 small things:
* let's keep the braces (`if {} else {}`) for clarity
* I find that `if (error) {...} else {...}` reads better? Shrugs.

File: contracts/Blockbin.sol [hunk 6]
```diff
@@ -15,6 +15,10 @@ contract Blockbin {
         bytes data;
     }
 
+    function Blockbin() {
```
Comment (ArnaudBrousseau) [inline contracts/Blockbin.sol:18]: Why is that needed?

**Review by ArnaudBrousseau — COMMENTED** [review]
Nice! Thanks for putting this diff together <3

I've got a few comments but I'd be fine merging and iterating since a lot of this most likely will change in the near future (I'm finally settled down in France, with potentially more time for focused coding!)

File: blockbin/src/components/InputForm.js (hunk 2 above)
Comment (amhunt) [reply blockbin/src/components/InputForm.js]: yeah you're totally right. This only works locally. I've also been thinking about using https://github.com/coopermaruyama/react-web3 - do you think this is a good idea? Wrapping the page in their `Web3Provider` component?

File: contracts/Blockbin.sol (hunk 6 above)
Comment (amhunt) [reply contracts/Blockbin.sol:18]: I believe it was needed when creating the contract object using web3. I added this a while back though, so it might now be necessary (I added it when playing around in the console). I'll try deploying and instantiating everything without this constructor and see if it still works.

**Comment by amhunt** [comment]
@ArnaudBrousseau thanks for your comments! I'll address them tonight and merge.

File: blockbin/src/components/InputForm.js (hunk 4 above)
Comment (amhunt) [reply blockbin/src/components/InputForm.js]: I'm not sure what should go here so I just put a really high gas limit. Should we be using the gas estimation utils in `web3` maybe?

File: blockbin/src/components/InputForm.js (hunk 2 above)
Comment (ArnaudBrousseau) [reply blockbin/src/components/InputForm.js]: I was just saying that right now I wouldn't even bother accommodating for users without metamask. A simple `alert` will do for now. react-web3 would be good to integrate in a separate PR :)

File: blockbin/src/components/InputForm.js (hunk 4 above)
Comment (ArnaudBrousseau) [reply blockbin/src/components/InputForm.js]: We also have plans to display the cost in gaz/dollars right next to the textarea so we'll have to do estimation at some point. If it's easy to get a rough estimate with `web3` go for it! (otherwise leave a comment and somebody else can pick this up in a separate PR)

**Comment by amhunt** [comment]
@ArnaudBrousseau I left two "TODO"s and merged. Also, could you look at the comment I made about the two callbacks? Not really sure of the best way to deal with these double callback situations.

File: blockbin/src/components/InputForm.js [hunk 7]
```diff
@@ -1,16 +1,95 @@
...
+    }
+
+    this.contractInstance.dumpCube.estimateGas(
+      cubeBytes, 
+      contentHash, 
+      { from: this.web3.eth.accounts[0] },
+      function(error, result){
+        if (!error) {
+          console.log('gas estimate: ' + result);
+          submitTx(result);
```
Comment (amhunt) [inline blockbin/src/components/InputForm.js:74]: @ArnaudBrousseau I added this function (defined immediately above). what's the best way to deal with this ? Nesting the two calls felt wrong but idk what else to do when the result of `estimateGas` is necessary for the call to `sendTransaction`.

Comment (ArnaudBrousseau) [reply blockbin/src/components/InputForm.js:74]: @amhunt you can avoid nesting by naming and hoisting out the anonymous callback passed to `estimateGas`
```js
const submitTx = function(gasEstimate) {...};

const estimateGasCb = function(error, result) {
    if (!error) {
        submitTx(result);
    } else {...}
};

this.contractInstance.estimateGas(arg1, arg2, arg3, estimateGasCb);
    arg2,
    arg3,
    estimateGasCb,
)
```
The other option is to wrap this in a promise. But it'll get a bit messy:
```js
function estimateGasPromise(method, ...args){
    return new Promise(function(resolve, reject){
        method.estimateGas(args, function(err, result) {
            if (err !== null) {
                return reject(err);
            } else {
                resolve(result);
            }
        })
    });
}
// This lets you do:
estimateGasPromise(this.contractInstance, arg1, arg2, arg3).then(submitTx);
```

I've got a runnable example at https://repl.it/@arnaudbrousseau/CurlyJauntyHellbender to show what I mean.

Unless you feel strongly I wouldn't bother changing what you have now, seems readable and perfectly usable as-is! ;)
